File: core/src/main/scala/offheap/Sanitizer.java
package offheap;

import static offheap.internal.UnsafeHolder.UNSAFE;
import static offheap.internal.CheckedHolder.CHECKED;

public final class Sanitizer {
    static final long ID_MASK = Long.MAX_VALUE << 48;
    static final long ADDR_MASK = Long.MAX_VALUE >> 16;

    public static long pack(short id, long addr) {
        return ((long) id << 48) | addr;
    }
    public static short unpackId(long addr) {
        return (short) ((addr & ID_MASK) >> 48);
    }
    public static long unpackAddr(long addr) {
        return addr & ADDR_MASK;
    }

    public static long validate(long addr) throws InaccessibleMemoryException {
        if (CHECKED) {
            if (!accessible(unpackId(addr)))
                throw new InaccessibleMemoryException("");
            return unpackAddr(addr);
        } else {
            return addr;
        }
    }

    static final short MAX_ID = Short.MAX_VALUE;
    static final long arr = UNSAFE.allocateMemory(MAX_ID);
    static int last = 0;
    static int count = 0;

    static int advance(int last) {
        int inc = last + 1;
        if (inc < MAX_ID)
            return inc;
        else
            return 0;
    }

    public synchronized static short register() {
        while (UNSAFE.getByte(arr + last) != 0)
            last = advance(last);
        short res = (short) last;
        UNSAFE.putByteVolatile(null, arr + last, (byte) 1);
        last = advance(last);
        count += 1;
        return res;
    }

    public synchronized static void unregister(short id) {
        count -= 1;
        UNSAFE.putByteVolatile(null, arr + id, (byte) 0);
    }

    public static boolean accessible(short id) {
        return UNSAFE.getByteVolatile(null, arr + id) == 1;
    }
}

File: core/src/main/scala/offheap/internal/Sanitizer.scala
package scala.offheap
package internal

import java.{lang => jl}
import internal.SunMisc.UNSAFE

object Sanitizer {
  private[this] final val UNPACKED_ID_MASK = 65535L
  private[this] final val ID_MASK = jl.Long.MAX_VALUE << 48
  private[this] final val ADDR_MASK = jl.Long.MAX_VALUE >> 16

  def pack(id: Long, addr: Addr): Addr = (id << 48) | addr
  def unpackId(addr: Addr): Long = (addr >> 48) & UNPACKED_ID_MASK
  def unpackAddr(addr: Addr): Addr = addr & ADDR_MASK

  private[this] final val MAX = 65536
  private[this] val valid = new java.util.concurrent.atomic.AtomicIntegerArray(MAX)
  private[this] val trans = new java.util.concurrent.atomic.AtomicLong(0L)

  private def truncate(v: Int): Int =
    if (v == 0) 1 else v
  private def advance(prev: Int): Int = {
    val inc = prev + 1
    if (inc < MAX) inc
    else 1
  }

  def register(): Long = {
    var commit = false
    var res = 0
    do {
      val prevtrans = trans.get
      val start = truncate((prevtrans % MAX).toInt)
      res = advance(start)
      while (valid.get(res) == 1 && res != start)
        res = advance(res)
      if (res == start && trans.get == prevtrans)
        throw new IllegalArgumentException(
          s"can't have more than ${MAX-1} regions open in checked memory mode")
      commit = valid.compareAndSet(res, 0, 1)
      trans.incrementAndGet
    } while (!commit)
    res.toLong
  }

  def unregister(id: Long): Unit = {
    valid.compareAndSet(id.toInt, 1, 0)
    trans.incrementAndGet
  }

  def validate(addr: Addr): Addr =
    if (Checked.MEMORY) {
      val id = unpackId(addr).toInt
      if (id != 0 && valid.get(id) != 1) {
        throw new InaccessibleMemoryException
      }
      unpackAddr(addr)
    } else {
      addr
    }
}
